hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/utils/Pipelines.java

```diff
@@ -316,9 +318,8 @@ public static DataStream<HoodieRecord> rowDataToHoodieRecord(Configuration conf,
   public static DataStream<Object> hoodieStreamWrite(Configuration conf, DataStream<HoodieRecord> dataStream) {
     if (OptionsResolver.isBucketIndexType(conf)) {
       WriteOperatorFactory<HoodieRecord> operatorFactory = BucketStreamWriteOperator.getFactory(conf);
-      int bucketNum = conf.getInteger(FlinkOptions.BUCKET_INDEX_NUM_BUCKETS);
-      String indexKeyFields = conf.getString(FlinkOptions.INDEX_KEY_FIELD);
-      BucketIndexPartitioner<HoodieKey> partitioner = new BucketIndexPartitioner<>(bucketNum, indexKeyFields);
+      dataStream = addBucketBootstrapIfNecessary(conf, dataStream);
+      Partitioner<HoodieKey> partitioner = BucketIndexPartitioner.instance(conf);
```

Comment:
Can we avoid the sub-pipeline for buckets metadata initializing and move them into the `ConsistentBucketStreamWriteFunction` ?

Reply:
We use the sub-pipeline to have a singleton to handle concurrent metadata initialization. Putting the initialization to StreamWriteFunction would requires the StreamCoordinator to handle special events (other than the current `WriteMetadataEvent`).

Reply:
Shouldn't the initialization of metadata under each partition be multi-writer safe ?

Reply:
Unfortunately not... Multi-writer safe is only possible if the underlying file system provides atomic-rename guarantee. For local file system or object store like OSS, multiple writers may contend to create the initial hashing metadata (with name 000000.hashing_metadata), causing inconsistent metadata states of writers.

ps. the metadata is partition-wide.

Reply:
can we use a file lock here to guard the consistency ? We already have impl in master as `FileSystemBasedLockProvider` ?

Reply:
Another way is to have a constant template for initial hashing metadata, e.g., fixed file group uuid. So that contention is ok since writers are creating metadata with the same content. Like how we handle `.hoodie_partition_metadata`.

Reply:
We can have a try.